File: examples/perl/subscribe_subject_filter.pl
#!/usr/bin/env perl

# Copyright 2007-2023 United States Government as represented by the
# Administrator of The National Aeronautics and Space Administration.
# No copyright is claimed in the United States under Title 17, U.S. Code.
# All Rights Reserved.

#
# @file subscribe_subject_filter.pl
# 
# This file contains a simple example outlining how to subscribe
# to and retrieve messages from the middleware bus while using subject
# exclusion filtering to reduce the scope of a wildcarded subject subscription.
#
# This example program is intended to be run before starting the
# 'publish_subject_filter' example program.
#


use strict;
use libgmsec_perl;

*isa = \&UNIVERSAL::isa;


sub main
{
    # Load the command-line input into a GMSEC Config object
    my $config = undef;
    if ($#ARGV ge 0)
    {
        $config = libgmsec_perl::Config::create(\@ARGV);
    }
    else
    {
        $config = libgmsec_perl::Config->new();
    }

    # Initialize log level for output
    initializeLogLevel($config);

    # Deduce subscription topic for the example program
    my $ALL_MESSAGES_TOPIC = "";
    my $HB_MESSAGES_TOPIC  = "";
    my $level = $config->getIntegerValue("gmsec-schema-level", $libgmsec_perl::Specification::SchemaLevel_LEVEL_0);

    if ($level == $libgmsec_perl::Specification::SchemaLevel_LEVEL_0)
    {
        $ALL_MESSAGES_TOPIC = "C2MS.>";
        $HB_MESSAGES_TOPIC  = "C2MS.*.*.*.*.*.MSG.HB.+";
    }
    elsif ($level == $libgmsec_perl::Specification::SchemaLevel_LEVEL_1)
    {
        $ALL_MESSAGES_TOPIC = "C2MS-PIPE.>";
        $HB_MESSAGES_TOPIC  = "C2MS-PIPE.*.*.*.*.*.MSG.HB.+";
    }
    else
    {
        $ALL_MESSAGES_TOPIC = "GMSEC.>";
        $HB_MESSAGES_TOPIC  = "GMSEC.*.*.MSG.HB.+";
    }

    eval
    {
        # Create connection instance.
        my $conn = libgmsec_perl::Connection::create( $config );

        # Establish connection to the GMSEC Bus.
        $conn->connect();
		
        # Output information regarding the API version and connection
        libgmsec_perl::logInfo(libgmsec_perl::Connection::getAPIVersion());
        libgmsec_perl::logInfo("Middleware version = " . $conn->getLibraryVersion());

        # Subscribe to receive all messages using the GMSEC specification
        libgmsec_perl::logInfo("Subscribe to: " . $ALL_MESSAGES_TOPIC);
        $conn->subscribe( $ALL_MESSAGES_TOPIC );

        # Add specific message topic to the Connection's exclusion filter
        libgmsec_perl::logInfo("Filtering out: " . $HB_MESSAGES_TOPIC);
        $conn->excludeSubject( $HB_MESSAGES_TOPIC );

        # Receive first message (LOG) sent by publisher
        libgmsec_perl::logInfo("Waiting to receive a message...");
        my $rcvdMsg = $conn->receive();

        libgmsec_perl::logInfo("Received LOG message:\n" . $rcvdMsg->toXML());
        $rcvdMsg->acknowledge();
        libgmsec_perl::Message::destroy( $rcvdMsg );

        # Although a second message (HB) is sent by the publisher, it will not
        # be delivered to this Connection object
        libgmsec_perl::logInfo("Waiting to receive other message(s)...");
        $rcvdMsg = $conn->receive(5000);

        if (defined $rcvdMsg)
        {
            libgmsec_perl::log_warning("Received unexpected message:\n" . $rcvdMsg->toXML());
            $rcvdMsg->acknowledge();
            libgmsec_perl::Message::destroy($rcvdMsg);
        }
        else
        {
            libgmsec_perl::logInfo("As expected, a timeout occurred (i.e. no HB message received)");
        }

        # Disconnect from the GMSEC Bus.
        $conn->disconnect();

        # Destroy the connection instance
        libgmsec_perl::Connection::destroy($conn);
    };
    if (isa($@, 'libgmsec_perl::GmsecException'))
    {
        # If here, the most likely case is the middleware could not be
        # deduced, or a failure occurred when attempting to connect to
        # the GMSEC Bus.
        my $error = $@;
        libgmsec_perl::logError("Exception: " . $error->what());
        return -1;
    }

    return 0;
}


sub initializeLogLevel
{
    my ($config) = @_;
    my $level = libgmsec_perl::Log::fromString( $config->getValue("loglevel", "info") );
    libgmsec_perl::Log::setReportingLevel($level);
}


# Main entry point
#
exit( main() );

File: examples/python3/subscribe_subject_filter.py
#!/usr/bin/env python3

# Copyright 2007-2025 United States Government as represented by the
# Administrator of The National Aeronautics and Space Administration.
# No copyright is claimed in the United States under Title 17, U.S. Code.
# All Rights Reserved.


"""
 @file subscribe_subject_filter.py
 
 This file contains a simple example outlining how to subscribe
 to and retrieve messages from the middleware bus while using subject
 exclusion filtering to reduce the scope of a wildcarded subject subscription.

 This example program is intended to be run before starting the
 'publish_subject_filter' example program.
"""

import libgmsec_python3 as lp
import sys


class subscribe_subject_filter:
    def __init__(self):
        # Set up connection configuration options using values from the command line.
        self.config = lp.Config(sys.argv)

        # Initialize log level for output
        self.initialize_log_level()

        # Deduce subscription topics for the example program
        self.ALL_MESSAGES_TOPIC = ""
        self.HB_MESSAGES_TOPIC  = ""
        level = self.config.get_integer_value("gmsec-schema-level", lp.Specification.SchemaLevel_LEVEL_0)

        if level == lp.Specification.SchemaLevel_LEVEL_0:
            self.ALL_MESSAGES_TOPIC = "C2MS.>"
            self.HB_MESSAGES_TOPIC  = "C2MS.*.*.*.*.*.MSG.HB.+"

        elif level == lp.Specification.SchemaLevel_LEVEL_1:
            self.ALL_MESSAGES_TOPIC = "C2MS-PIPE.>"
            self.HB_MESSAGES_TOPIC  = "C2MS-PIPE.*.*.*.*.*.MSG.HB.+"

        else:
            self.ALL_MESSAGES_TOPIC = "GMSEC.>"
            self.HB_MESSAGES_TOPIC  = "GMSEC.*.*.MSG.HB.+"


    def initialize_log_level(self):
        level = lp.Log.from_string( self.config.get_value("loglevel", "info") )
        lp.Log.set_reporting_level(level)


    def run(self):
        try:
            # Create connection instance.
            conn = lp.Connection(self.config)

            # Establish connection to the GMSEC Bus.
            conn.connect()
		
            # Output information regarding the API version and connection
            lp.log_info(lp.Connection.get_api_version())
            lp.log_info("Middleware version = " + conn.get_library_version())

            # Subscribe to receive all messages using the GMSEC specification
            lp.log_info("Subscribe to: " + self.ALL_MESSAGES_TOPIC)
            conn.subscribe(self.ALL_MESSAGES_TOPIC)

            # Add specific message topic to the Connection's exclusion filter
            lp.log_info("Filtering out: " + self.HB_MESSAGES_TOPIC)
            conn.exclude_subject(self.HB_MESSAGES_TOPIC)

            # Receive first message (LOG) sent by publisher
            lp.log_info("Waiting to receive a message...")
            rcvdMsg = conn.receive()

            lp.log_info("Received LOG message:\n" + rcvdMsg.to_xml())
            rcvdMsg.acknowledge()
            lp.Message.destroy(rcvdMsg)

            # Although a second message (HB) is sent by the publisher, it will not
            # be delivered to this Connection object
            lp.log_info("Waiting to receive other message(s)...")
            rcvdMsg = conn.receive(5000)

            if rcvdMsg == None:
                lp.log_info("As expected, a timeout occurred (i.e. no HB message received)")
            else:
                lp.log_warning("Received unexpected message:\n" + rcvdMsg.to_xml())
                rcvdMsg.acknowledge()
                lp.Message.destroy(rcvdMsg)

            # Disconnect from the GMSEC Bus.
            conn.disconnect()

            # Destroy the Connection
            del conn

        except lp.GmsecError as e:
            # If here, the most likely case is the middleware could not be
            # deduced, or a failure occurred when attempting to connect to
            # the GMSEC Bus.
            lp.log_error("Exception: " + str(e))


# Main entry point
ex = subscribe_subject_filter()
ex.run()
